### PR TITLE
EC: Reimplement EVP_PKEY_CTX_set_ec_param_enc() to support providers

### LINKED ISSUE #12852: [3.0.0] EVP_PKEY_CTX_set_ec_param_enc() is not supported
>> @danbev [said](https://github.com/openssl/openssl/issues/12102#issuecomment-653438152)
>>
>> I'm currently working on updating Node.js to build and run against OpenSSL 3.x and we use EVP_PKEY_CTX_set_ec_param_enc. (I've extracted this into a standalone program, ec.c which might be easier to follow)
>>
>> This is causing test failures for us at the moment, but we can add checks to skip them until a fix lands upstream. I just wanted to let you know that we've run into this, and please let us know if there is anything we should do differently.

> @romen [said](https://github.com/openssl/openssl/issues/12102#issuecomment-653457021)
>
> @danbev with named curve it should be working, what failure do you see?

@romen There is no OpenSSL failure/error reported as far as I can tell, below is the output of the standalone program `ec` and the error in Node.js. 

What I see happening is that the function [legacy_ctrl_to_param](https://github.com/openssl/openssl/blob/94fcec0902ec6df54a6286f61e35fa6f93815fd6/crypto/evp/pmeth_lib.c#L808) is called and since there is no case that matches [EVP_PKEY_EC](https://github.com/openssl/openssl/blob/94fcec0902ec6df54a6286f61e35fa6f93815fd6/crypto/evp/pmeth_lib.c#L850) the [return statement](https://github.com/openssl/openssl/blob/94fcec0902ec6df54a6286f61e35fa6f93815fd6/crypto/evp/pmeth_lib.c#L938) will be reached and 0 will be returned. This is then checked in Node.js as interpreted as an error, see details below:

<details>
<summary>ec output</summary>

```console
$ ./ec 
Elliptic Curve example
curve_nid of secp256k1: 714
EVP_PKEY_CTX_set_ec_param_enc retuned: 0
EVP_PKEY_CTX_set_ec_param_enc failed
errno: 0, error:00000000:lib(0)::reason(0)

```
</details>
<details>
 <summary>Node.js error</summary>

```console
$ ./out/Debug/node test/parallel/test-crypto-keygen.js
./out/Debug/node[876391]: ../src/node_crypto.cc:6350:void node::crypto::GenerateKeyPairJob::ToResult(v8::Local<v8::Value>*, v8::Local<v8::Value>*, v8::Local<v8::Value>*): Assertion `!errors_.empty()' failed.
 1: 0xe4f878 node::DumpBacktrace(_IO_FILE*) [./out/Debug/node]
 2: 0xf07561 node::Abort() [./out/Debug/node]
 3: 0xf07617  [./out/Debug/node]
 4: 0x10d636d node::crypto::GenerateKeyPairJob::ToResult(v8::Local<v8::Value>*, v8::Local<v8::Value>*, v8::Local<v8::Value>*) [./out/Debug/node]
 5: 0x10d620e node::crypto::GenerateKeyPairJob::AfterThreadPoolWork() [./out/Debug/node]
 6: 0x10d40a5 node::crypto::CryptoJob::AfterThreadPoolWork(int) [./out/Debug/node]
 7: 0xec1b1c node::ThreadPoolWork::ScheduleWork()::{lambda(uv_work_s*, int)#2}::operator()(uv_work_s*, int) const [./out/Debug/node]
 8: 0xec1b42 node::ThreadPoolWork::ScheduleWork()::{lambda(uv_work_s*, int)#2}::_FUN(uv_work_s*, int) [./out/Debug/node]
 9: 0x1d6dabc  [./out/Debug/node]
10: 0x1d6da01  [./out/Debug/node]
11: 0x1d72249  [./out/Debug/node]
12: 0x1d898a8  [./out/Debug/node]
13: 0x1d72bca uv_run [./out/Debug/node]
14: 0xf752a1 node::NodeMainInstance::Run() [./out/Debug/node]
15: 0xeb4e45 node::Start(int, char**) [./out/Debug/node]
16: 0x23331b2 main [./out/Debug/node]
17: 0x7f319142d1a3 __libc_start_main [/lib64/libc.so.6]
18: 0xdfecee _start [./out/Debug/node]
Aborted (core dumped)
```
</details>

### PR DESCRIPTION
**Pull request by levitte**
Fixes #12852

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
I've grepped through our test/ and apps/ source, and it seems that this control isn't used anywhere that we can test, so I would appreciate it if someone helped me create a sensible test.  #12852 mentions an `ec.c`, which might be helpful for local testing at least, but we might want to consider adding use of this control somewhere among already existing test programs.  @romen?

**Comment by slontis** [comment]
@romen gave you the test :).. I am adding it now.

**Comment by levitte** [comment]
> @romen gave you the test :)

????? I must be blind, I still can't see where.  But hey, if you're adding it, that's great!  Thanks :-)

**Comment by slontis** [comment]
@levitte changes are in now..
The last commit is related to macro references in the pod file. They are all functions now.. Feel free to modify it if I have messed anything up.

**Comment by levitte** [comment]
I see the added test...  have you checked that it actually hits `EVP_PKEY_CTX_set_ec_param_enc`?  As far as I know the code, the string based controls are directly converted to OSSL_PARAM arrays for provider-native keys, so should completely bypass the functionality that's modified in this PR...  We need a test with an explicit `EVP_PKEY_CTX_set_ec_param_enc` call...

**Comment by levitte** [comment]
I've added a small squash commit for the general documentation fixup...  some text duplication found itself in the HISTORY section

**Comment by levitte** [comment]
Side note: one might wonder why `EVP_PKEY_CTX_set_ec_param_enc` wasn't name `EVP_PKEY_CTX_set_ec_paramgen_enc`, *like all other paramgen type oof parameter setting functions*!

**Comment by levitte** [comment]
I'm writing a test for `test/evp_extra_test.c`...

**Comment by levitte** [comment]
I believe this PR is complete at this point.  Time for final reviews!

**Review by slontis — APPROVED** [review]
LGTM

File: test/recipes/15-test_genec.t [hunk 1]
```diff
@@ -197,6 +198,20 @@ ok(run(app([ 'openssl', 'genpkey',
...
+             '-text'])),
+   "generate a key with explicit encoding");
+
+ok(run(app([ 'openssl', 'genpkey',
+             '-algorithm', 'EC',
+             '-pkeyopt', 'ec_paramgen_curve:P-256',
+             '-pkeyopt', 'ec_param_enc:named_curve',
+             '-text'])),
+   "generate a key with named_curve encoding");
+
```
Comment (romen) [inline test/recipes/15-test_genec.t:214]: Nit: These 2 tests are a bit superfluous: just after them we test again the same thing for all supported curves!

**Review by romen — APPROVED** [review]
LGTM.

I have 1 nit to report, but I believe that if @levitte decides to remove the duplicate tests this shouldn't really affect the approval from @slontis.

**Comment by romen** [comment]
> I'm writing a test for `test/evp_extra_test.c`...

@levitte the way you tested it with the export callback looks very cool!
I don't think I could have done it like this and I was struggling to find a proper way to check on the produced serialization output. This looks way cleaner!

**Comment by romen** [comment]
AppVeyor's error is relevant:

```
evp_extra_test-bin-evp_extra_test.obj : error LNK2019: unresolved external symbol strcasecmp referenced in function ec_export_get_encoding_cb
```

File: test/recipes/15-test_genec.t (hunk 1 above)
Comment (levitte) [reply test/recipes/15-test_genec.t:214]: Good point.  I'll remove this, then

Comment (levitte) [reply test/recipes/15-test_genec.t:214]: Done

**Comment by levitte** [comment]
> AppVeyor's error is relevant:

Fixed

**Review by romen — APPROVED** [review]
Formal reapproval. 

I think both changes pushed after approval are trivial enough that a new grace period is not required.

**Comment by levitte** [comment]
Merged

7229a2f4ab9b4d8cecf44be58adeb14e195ff051 EC: Reimplement EVP_PKEY_CTX_set_ec_param_enc() to support providers
33dd828d976fc6b1b508f090b8fd91fbf9ab7cce Update doc for EVP_PKEY_CTX_set_ec_param_enc()
96bb4ff9b8473d01d9ac9301ec0f796898c8f459 Fix EVP_PKEY_CTX_ctrl() documentation
655f73cecf411737cef9debdfa4c0b8b041656df EVP: Add the internal convenience function evp_keymgmt_util_export()
f0c62c53286b3a79cc059354a6d3425b35bb1d50 TEST: Add a test of EC key generation with encoding spec

**Comment by levitte** [comment]
@slontis, thanks so much for the added docs, and for trying to help with the test!

**Pull request closed without merging**